### PR TITLE
Feature/testing

### PR DESCRIPTION
**Pull request by rna**
# Hi, Good Day

- Required tests are implemented using Jest framework.

Thanks.

### CODE REVIEW

**Review by stratospherique — CHANGES_REQUESTED** [review]
### Great start @team :clap: However, some requirements need to be fulfilled in order to get approval for this milestone. Check :arrow_down: 

##### General (Required) :red_circle: 
* [ ] Pull request should have passing code linter checks.
##### Tests implementation
- The unit & their tests are working from a logic stand point. However, they need to fulfill the fowwing requirements.
#### `capitalize` ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/javascript/testing_practice/capitalize_requirement.png) from [Odin](https://www.theodinproject.com/courses/javascript/lessons/testing-practice))

* [ ] `capitalize` is a function that takes a string as an argument and returns that string with the first character capitalized
* [ ] There is a Jest test for the `capitalize` function and it's passing

#### `reverseString` ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/javascript/testing_practice/reverse_requirement.png) from [Odin](https://www.theodinproject.com/courses/javascript/lessons/testing-practice))

* [ ] `reverseString` is a function that takes a string as an argument and returns that string reversed
* [ ] There is a Jest test for the `reverseString` function and it's passing

#### `calculator` ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/javascript/testing_practice/calculator_requirement.png) from [Odin](https://www.theodinproject.com/courses/javascript/lessons/testing-practice))

* [ ] It is an object
* [ ] It has an `add` method
* [ ] It has a `subtract`
* [ ] It has a `multiply` method
* [ ] It has a `divide` method
* [ ] There is a Jest test for the `calculator` object that tests the `add`, `subtract`, `multiply` and `divide` methods

#### Caesar Cipher ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/javascript/testing_practice/caesar_requirement.png) from [Odin](https://www.theodinproject.com/courses/javascript/lessons/testing-practice))

* [ ] Caesar Cipher is implemented
* [ ] It has a test for wrapping from `z` to `a`
* [ ] It has a test for keeping the same case
* [ ] It has a test for punctuation

#### Array Analysis ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/javascript/testing_practice/array_analysis_requirement.png) from [Odin](https://www.theodinproject.com/courses/javascript/lessons/testing-practice))

* [ ] It has a function that takes an array of numbers as an argument and returns an object
* [ ] The returned object has the `average` of the numbers in the array as a property
* [ ] The returned object has the `min` of the numbers in the array as a property
* [ ] The returned object has the `max` of the numbers in the array as a property
* [ ] The returned object has the `length` of the numbers in the array as a property
* [ ] There is a Jest test for function to calculate correctly `average`, `min`, `max` and `length`

### That's it :relieved: When finished, kindly submit a code re-review request :pray: 
**_Ahmed Mahfoudh_**

File: src/jsfiles/caesar.js [hunk 1]
```diff
@@ -0,0 +1,14 @@
+const caesar = (str) => {
```
Comment (mcrd25) [inline src/jsfiles/caesar.js]: You are missing the shift parameter. Remember Caesar Cipher is not only for shift of 1 but can be any. This technically not a full caesar cipher. 
Reread here: http://practicalcryptography.com/ciphers/caesar-cipher/
Implement changes accordingly and update tests accordingly.

**Review by mcrd25 — CHANGES_REQUESTED** [review]
# Changes Required :mag: 

Hi Team! :wave: 

Like what I see so far! :smile: :+1: 

The only issue is that the cipher is not fully implemented. Once you address this, you should get  :+1: 

***Please makes changes and submit another review request.
If you have any questions, let me know here or on slack and I will be glad to clarify with you. You can also reach me and connect with me on my social media or email.***

*Happy coding 😄,*
**[--Maya](https://github.com/mcrd25)**
[GitHub](https://github.com/mcrd25) [Twitter](https://twitter.com/mcrd25) [linkedin](https://www.linkedin.com/in/mayadouglas/) [GitLab](https://gitlab.com/mcrd25)

---
*P.S. all requested changes not marked by [OPTIONAL] or are not positive comments are **REQUIRED***

File: src/jsfiles/arrayAnalysis.js [hunk 2]
```diff
@@ -0,0 +1,11 @@
+const analyze = (arr) => {
```
Comment (uryelah) [inline src/jsfiles/arrayAnalysis.js:1]: - [ ] [OPTIONAL] What happens if you use this method with an empty array? And if it's not even an array? Test it and try to add some edge cases handling in case something breaks.

File: src/jsfiles/caesar.js [hunk 3]
```diff
@@ -0,0 +1,12 @@
+const encode = (e, shift) => {
+  if (/^[a-zA-Z]/.test(e)) {
```
Comment (uryelah) [inline src/jsfiles/caesar.js:2]: - [ ] [OPTIONAL] How would you implement this so it also allows for numbers or special characters(?!@#...etc) in your strings?

File: src/jsfiles/calculator.js [hunk 4]
```diff
@@ -0,0 +1,8 @@
+const calculator = {
+  add: (a, b) => a + b,
+  subtract: (a, b) => a - b,
+  divide: (a, b) => a / b,
```
Comment (uryelah) [inline src/jsfiles/calculator.js:4]: - [ ] [OPTIONAL] What if `b` is 0?

File: src/test/capitalize.test.js [hunk 5]
```diff
@@ -0,0 +1,5 @@
+import capitalize from '../jsfiles/capitalize';
+
+test('capitalize first Letter', () => {
+  expect(capitalize('here')).toBe('Here');
```
Comment (uryelah) [inline src/test/capitalize.test.js:4]: - [x] [For next projects] When you write tests for real remember to add more than one case for each method, otherwise it could only be passing with one specific input and you want to make sure that doesn't happen.

File: src/test/reverse.test.js [hunk 6]
```diff
@@ -0,0 +1,5 @@
+import reverse from '../jsfiles/reverse';
+
+test('Return reversed string', () => {
+  expect(reverse('backwards')).toBe('sdrawkcab');
```
Comment (uryelah) [inline src/test/reverse.test.js:4]: - [x] [For next projects] When you write tests for real remember to add more than one case for each method, otherwise it could only be passing with one specific input and you want to make sure that doesn't happen.

**Review by uryelah — APPROVED** [review]
Hi **Ramesh Allu & Tenny Zvaita**, I'm Sarah your reviewer for this PR 👩‍🚀 

Good job with the completing the Cesar Cyoher function and updating the testing for it!

Go forth and write some rocking tests in your next project!  👨‍🎤   🎸  

-------

## **Your project is approved! 🎊 🥅 💛**

**You may close this Pull Request and merge the changes**


---------

    
    
**Keep being awesome devs and good luck with your next project!**

---------

Cheers and Happy coding!👏👏👏

Code review by **Sarah Uryelah Silva Chamorro**

(Feel free to contact me on slack by the handle of  **sarah** if you have any additional question)

**Comment by rna** [comment]
## Thanks for the detailed review @uryelah :+1: 
#### This is what I expect from a TSE. You nailed it. :clap: :clap: :fireworks:  Keep up the good work. :tada:
